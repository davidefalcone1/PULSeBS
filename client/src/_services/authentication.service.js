import { BehaviorSubject } from 'rxjs';


const currentUserSubject = new BehaviorSubject(JSON.parse(localStorage.getItem('user')));

export const authenticationService = {
    login,
    logout,
    user: currentUserSubject.asObservable(),
    get currentUserValue () { return currentUserSubject.value }
};

function handleResponse(response) {
    return response.text().then(text => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if ([401, 403].indexOf(response.status) !== -1) {
                // auto logout if 401 Unauthorized or 403 Forbidden response returned from api
                authenticationService.logout();
            }

            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }

        return data;
    });
}

function login(username, password) {
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ username, password })
    };
    return new Promise(async function (resolve, reject) {
        try {
            const response = await fetch('/users/authenticate', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({ username, password }),
            });
            const user = await response.json().then(handleResponse);
            if (response.ok) {
                localStorage.setItem('user', JSON.stringify(user));
                currentUserSubject.next(user);
                resolve(user);
            }
            else
                reject();
        }
        catch (e) {
            reject();
        }
    });

}
function logout() {

    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
    };
    localStorage.removeItem('user');
    currentUserSubject.next(null);
    return fetch('/logout',requestOptions);
}