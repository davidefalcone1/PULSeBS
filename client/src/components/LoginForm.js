import React, { Fragment } from 'react';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';
import { Redirect } from 'react-router-dom';
import { AuthContext } from '../_services/AuthContext'

class LoginForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = { username: '', password: '', submitted: false };
    }

    onChangeUsername = (event) => {
        this.setState({ username: event.target.value });
    };

    onChangePassword = (event) => {
        this.setState({ password: event.target.value });
    };

    handleSubmit = (event, onLogin) => {
        event.preventDefault();
        onLogin(this.state.username, this.state.password).then(() => {
        });
        this.setState({ submitted: true });
    }

    render() {
        // if (this.state.submitted)
        //     return <Redirect to='/' />;
        return (
            <AuthContext.Consumer>
                {(context) => (
                    <>
                        {!context.user && 
                            <Fragment>
                                <div id='form' className="col-sm-4 mx-auto">
                                    <div className="card">
                                        <article className="card-group-item">
                                            <header className="card-header">
                                                <h5 className="title"><em id ='form-title'>Insert your credentials </em></h5>
                                            </header>  
                                            <div className="filter-content">
                                                <div className="card-body">
                                                    <div className="table-responsive">

                                                        <Form method="POST" onSubmit={(event) => this.handleSubmit(event, context.loginUser)}>
                                                            <Form.Group controlId="username">
                                                                <Form.Label>E-mail</Form.Label>
                                                                <Form.Control className="inputEmail" type="email" name="email" placeholder="E-mail" value={this.state.username} onChange={(ev) => this.onChangeUsername(ev)} required autoFocus />
                                                                <small className="form-text text-muted">Email is required</small>
                                                            </Form.Group>



                                                            <Form.Group controlId="password">
                                                                <Form.Label>Password</Form.Label>
                                                                <Form.Control type="password" name="password" placeholder="Password" value={this.state.password} onChange={(ev) => this.onChangePassword(ev)} required />
                                                                <small className="form-text text-muted">Password is required</small>
                                                            </Form.Group>

                                                            <Button variant="primary" type="submit">Login</Button>
                                                        </Form>

                                                        {context.loginError &&
                                                            <Alert variant="danger">
                                                                {context.loginError}
                                                            </Alert>
                                                        }
                                                    </div>
                                                </div>
                                            </div>
                                        </article>
                                    </div>
                                </div>
                            </Fragment>
                        }
                        {context.configurationCompleted && context.isStudent && <Redirect to="lessonslist"/> }
                        {context.configurationCompleted && context.isTeacher && <Redirect to="myCoursesLessonslist"/> }
                    </>
                )}
            </AuthContext.Consumer>

        );
    }


}

export default LoginForm;