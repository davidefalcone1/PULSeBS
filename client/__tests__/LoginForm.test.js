import React from 'react';
import renderer from 'react-test-renderer';
import { AuthContext } from '../src/_services/AuthContext';
import LoginForm from '../src/components/LoginForm';
import { BrowserRouter } from 'react-router-dom';

describe('LoginForm', ()=>{
    test('Renders correctly', ()=>{
        const component = renderer.create(
            <BrowserRouter>
                <AuthContext.Provider value={{user: undefined}}>
                    <LoginForm/>
                </AuthContext.Provider>
            </BrowserRouter>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});