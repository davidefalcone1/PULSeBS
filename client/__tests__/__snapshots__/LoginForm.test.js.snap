// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Renders correctly 1`] = `
<div
  className="col-sm-4 mx-auto"
  id="form"
>
  <div
    className="card"
  >
    <article
      className="card-group-item"
    >
      <header
        className="card-header"
      >
        <h5
          className="title"
        >
          <em
            id="form-title"
          >
            Insert your credentials 
          </em>
        </h5>
      </header>
      <div
        className="filter-content"
      >
        <div
          className="card-body"
        >
          <div
            className="table-responsive"
          >
            <form
              className=""
              method="POST"
              onSubmit={[Function]}
            >
              <div
                className="form-group"
              >
                <label
                  className="form-label"
                  htmlFor="username"
                >
                  E-mail
                </label>
                <input
                  autoFocus={true}
                  className="inputEmail form-control"
                  id="username"
                  name="email"
                  onChange={[Function]}
                  placeholder="E-mail"
                  required={true}
                  type="email"
                  value=""
                />
                <small
                  className="form-text text-muted"
                >
                  Email is required
                </small>
              </div>
              <div
                className="form-group"
              >
                <label
                  className="form-label"
                  htmlFor="password"
                >
                  Password
                </label>
                <input
                  className="form-control"
                  id="password"
                  name="password"
                  onChange={[Function]}
                  placeholder="Password"
                  required={true}
                  type="password"
                  value=""
                />
                <small
                  className="form-text text-muted"
                >
                  Password is required
                </small>
              </div>
              <button
                className="btn btn-primary"
                disabled={false}
                type="submit"
              >
                Login
              </button>
            </form>
          </div>
        </div>
      </div>
    </article>
  </div>
</div>
`;
